README.md
Form library for Flutter, offering controlled inputs, user-friendly validation and more.

## Features

Define your fields as usual:
```dart
class Form {
    final String username;
    final String password;
    final String repeatPassword;

    final bool wasEverSubmitted; // user tried to submit
    
    ...
}
```

Add field validation:
```dart
class Form {
  ...

  late final usernameField = username
      .validate((value) => isLength(value, 8), "Min 8 chars")
      .then((value) => isLength(value, 0, 16), "Max 16 chars")
      .then(isAlphanumeric, "Only numbers and letters");

  late final passwordField = password.validate(
    (value) => isLength(value, 8),
    "Min 8 chars",
  );

  late final repeatPasswordField = repeatPassword.validate(
    (value) => value == password,
    "Passwords should match",
  );
}
```

Wrap your Flutter `TextField`s:
```dart
Widget build(BuildContext context) {
  final form = ... // retrieve your form
  return Column(
    children: [
      form.usernameField.builder(
        builder: (context, controller, errorText) => TextField(
          controller: controller, // provide controller
          decoration: InputDecoration(errorText: errorText),
          onChanged: formViewModel.onUsernameChanged, // formViewModel - is your state managemnt instance. It could be ChangeNotifier, Cubit e.t.c.
        ),
      ),
      form.passwordField.builder(
        builder: (context, controller, errorText) => TextField(
          controller: controller,
          decoration: InputDecoration(errorText: errorText),
          onChanged: formViewModel.onPasswordChanged,
        ),
      ),
      form.repeatPasswordField.builder(
        builder: (context, controller, errorText) => TextField(
          controller: controller,
          decoration: InputDecoration(errorText: errorText),
          onChanged: formViewModel.onRepeatPasswordChanged,
        ),
      ),
    ]
  );
}
```
**Now it just works**!

## Adjust moment errors appear

It's not very practical and user-friendly to show errors as user types.
Let's change it with `ReformScope`!

Wrap your widget tree with `ReformScope` like that:

```dart
Widget build(BuildContext context) {
  final form = ... // retrieve your form

  return ReformScope(
    shouldShowError: (field, fieldState) => form.wasEverSubmitted || fieldState.wasEverUnfocused;
    child: ... // your `Widget` with fields
  );
} 
```

Now error will become visible once:
* User submits your form. Since this moment it's appropriate to show errors as user types
* Field is unfocused after it was focused for the first time. It's appropriate to show user an error, once they finished typing.

## Handle submission

Let's have a quick look at `ChangeNotifier` submission implementation:

```dart
class FormViewModel extends ChangeNotifier {
  ...

  void submit() {
    this.form = form.copyWith(wasEverSubmitted: true);
    notifyListeners();

    // utility class to help with validation of multiple fields
    final isValid = Reform.validate([
      form.usernameField,
      form.passwordField,
      form.repeatPasswordField,
    ]);

    if (isValid) {
      // POST request to your server
    }
  }
}
```

## Immutability

Thought above example was based on immutabiliy of `Form`, it's completely fine to use `Reform` with mutable classes.
Just don't forget to replace `late final` with `get`ters.

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 2023-06-20

### Changes

---

Packages with breaking changes:

 - [`reform` - `v0.10.0`](#reform---v0100)

Packages with other changes:

 - [`reform_flutter` - `v0.7.0+5`](#reform_flutter---v0705)

Packages with dependency updates only:

> Packages listed below depend on other packages in this workspace that have had changes. Their versions have been incremented to bump the minimum dependency versions of the packages they depend upon in this project.

 - `reform_flutter` - `v0.7.0+5`

---

#### `reform` - `v0.10.0`

 - **BREAKING** **FEAT**: remove lazy evaluation.


## 2023-05-25

### Changes

---

Packages with breaking changes:

 - [`reform` - `v0.9.0`](#reform---v090)

Packages with other changes:

 - [`reform_flutter` - `v0.7.0+4`](#reform_flutter---v0704)

Packages with dependency updates only:

> Packages listed below depend on other packages in this workspace that have had changes. Their versions have been incremented to bump the minimum dependency versions of the packages they depend upon in this project.

 - `reform_flutter` - `v0.7.0+4`

---

#### `reform` - `v0.9.0`

 - **BREAKING** **FIX**: validateAsync signature.


## 2023-05-25

### Changes

---

Packages with breaking changes:

 - [`reform` - `v0.8.0`](#reform---v080)

Packages with other changes:

 - [`reform_flutter` - `v0.7.0+3`](#reform_flutter---v0703)

Packages with dependency updates only:

> Packages listed below depend on other packages in this workspace that have had changes. Their versions have been incremented to bump the minimum dependency versions of the packages they depend upon in this project.

 - `reform_flutter` - `v0.7.0+3`

---

#### `reform` - `v0.8.0`

 - **BREAKING** **FIX**: validateAsync signature.


## 2023-05-25

### Changes

---

Packages with breaking changes:

 - [`reform` - `v0.7.0`](#reform---v070)

Packages with other changes:

 - [`reform_flutter` - `v0.7.0+2`](#reform_flutter---v0702)

Packages with dependency updates only:

> Packages listed below depend on other packages in this workspace that have had changes. Their versions have been incremented to bump the minimum dependency versions of the packages they depend upon in this project.

 - `reform_flutter` - `v0.7.0+2`

---

#### `reform` - `v0.7.0`

 - **BREAKING** **FEAT**: async validation.


## 2023-05-25

### Changes

---

Packages with breaking changes:

 - [`reform` - `v0.6.0`](#reform---v060)

Packages with other changes:

 - [`reform_flutter` - `v0.7.0+1`](#reform_flutter---v0701)

Packages with dependency updates only:

> Packages listed below depend on other packages in this workspace that have had changes. Their versions have been incremented to bump the minimum dependency versions of the packages they depend upon in this project.

 - `reform_flutter` - `v0.7.0+1`

---

#### `reform` - `v0.6.0`

 - **BREAKING** **FEAT**: async validation.


## 2023-05-25

### Changes

---

Packages with breaking changes:

 - [`reform` - `v0.5.0`](#reform---v050)
 - [`reform_flutter` - `v0.7.0`](#reform_flutter---v070)

Packages with other changes:

 - There are no other changes in this release.

---

#### `reform` - `v0.5.0`

 - **BREAKING** **FEAT**: refactoring.

#### `reform_flutter` - `v0.7.0`

 - **BREAKING** **FEAT**: refactoring.


## 2023-05-02

### Changes

---

Packages with breaking changes:

 - [`reform` - `v0.4.0`](#reform---v040)
 - [`reform_flutter` - `v0.6.0`](#reform_flutter---v060)

Packages with other changes:

 - There are no other changes in this release.

---

#### `reform` - `v0.4.0`

 - **BREAKING** **FEAT**: refactoring.

#### `reform_flutter` - `v0.6.0`

 - **BREAKING** **FEAT**: refactoring.


## 2023-04-18

### Changes

---

Packages with breaking changes:

 - [`reform_flutter` - `v0.5.0`](#reform_flutter---v050)

Packages with other changes:

 - There are no other changes in this release.

---

#### `reform_flutter` - `v0.5.0`

 - **BREAKING** **FIX**: don't dispose controller, if it's passed through parameter.



example/lib/cubit/main_state.dart
part of 'main_cubit.dart';

@freezed
class MainState with _$MainState {
  factory MainState({
    required String username,
    required String password,
    required String repeatPassword,
    required int age,
    required bool isSubmittedOnce,
  }) = _MainState;

  MainState._();

  late final usernameField = username.validate((value) {
    if (value.length < 8) {
      return "Min 8 chars";
    } else if (value.length > 16) {
      return "Max 16 chars";
    } else if (isAlphanumeric(value)) {
      return "Only numbers and letters";
    }
    return null;
  });

  late final passwordField = password.validate(
    (value) => value.length < 8 ? "At least 8 characters" : null,
  );

  late final repeatPasswordField = repeatPassword.validate(
    (value) => value != password ? "Passwords should match" : null,
  );

  late final ageField = age.validate(
    (value) => value < 18 ? "You should be 18 y.o" : null,
  );
}
